Show torrent fields: creation_date, created_by, encoding

mario-nt opened the pull request:
Resolves #278.

Commit:
feat: [#278] show torrent fields creation date created by encoding

    Torrent details page now displays the creation date, created by and encoding fields.

mario-nt commented:
@josecelano Rebased and CI OK

josecelano commented on components/torrent/TorrentCreationDateTab.vue:
```diff
@@ -0,0 +1,52 @@
+  }
+});
+
+// Takes the date in milliseconds/Epoch Time and converts it to human readable format
```
Hi @mario-nt,

- I would move the comment to the function. If you think the name is not enough, you could change the name to `epochTimeToHumanReadableUTC` or `timestampToHumanReadableUTC`
- Timestamp is in seconds (`props.torrent.creation_date`), not milliseconds. The Date constructor accepts milliseconds (`new Date(date * 1000);`). We are passing seconds to the function.

josecelano commented on src/helpers/DateConverter.ts:
```diff
@@ -0,0 +1,4 @@
+export function epochToUtc (date: number) {
+  const convertedDate = new Date(date * 1000);
+  return convertedDate.toDateString();
+}
```
Hi @mario-nt since we are using TypeScript maybe we can introduce a new `Timestamp` type to make it explicit what the function needs.

josecelano replied:
I was using `snake_case` for JS functions and I guess also for filenames because they all start with a lowercase:

- `sanitizer.ts`
- `slug.ts`

josecelano replied:
It would be good to add a unit test for the function but [it's a pain to add unit tests with Nuxt](https://github.com/torrust/torrust-index-gui/issues/164).

josecelano replied:
> Hi @mario-nt since we are using TypeScript maybe we can introduce a new `Timestamp` type to make it explicit what the function needs.

Maybe something like:

```Typescript
type Timestamp = number;

export function epochToUtc (timestamp: Timestamp) {
  new Date(timestamp.toMilliseconds()).toDateString();
}
```

But you can even use a field and check that the number is an integer.

josecelano commented:
This is how the detail page looks with the new fields:

![image](https://github.com/torrust/torrust-index-gui/assets/58816/35782ce6-4c62-49ee-9155-cd2444bb64fd)

mario-nt committed:
refactor: [#297] Creation date now display as UTC

If the torrent creation date field is present in the torrent file, it is now shown as a UTC date
in a human readable format.

If there is no creation date, a no creation date provided message is displayed.

And if there is an error parsing the creation date or if it is a no valid date, an invalid date
text is shown.

mario-nt committed:
refactor: [#297] added error checks

Error checks added to the function that converts the torrent creation date from seconds to
UTC human readable format.

Refactor to the function logic's code has been applied.

mario-nt committed:
refactor: [#297] new error checks and refactoring

Date converter helper function now throws errors if the timestamp passed as argument is not an integer
and if the Date() constructor returns an invalid date.

mario-nt commented:
@josecelano Ready for review.

josecelano commented on components/torrent/TorrentCreationDateTab.vue:
```diff
@@ -0,0 +1,48 @@
+    <template v-if="!collapsed">
+      <div class="flex flex-col w-full h-full p-6 grow bg-base-100 rounded-2xl">
+        <template v-if="torrent.creation_date">
+          {{ unixTimeToHumanReadableUTC(torrent.creation_date) }}
```
Hi @mario-nt maybe inside the components we could use a wrapper helper with the name `formatDateFromTimestamp`.

mario-nt committed:
refactor: [#297] refactored error control

mario-nt commented:
@josecelano  Ready for review.

josecelano commented on src/helpers/DateConverter.ts:
```diff
@@ -0,0 +1,22 @@
+class InvalidDateError extends Error {}
+class WrongTimestamp extends Error {}
+
+// Takes the date in seconds from Epoch time and converts it to human readable format.
```
Hi @mario-nt I think we should use a format for comments like https://tsdoc.org/ but we can enforce that after [installing MegaLinter](https://github.com/torrust/torrust-index-gui/issues/366).

josecelano commented on src/helpers/DateConverter.ts:
```diff
@@ -0,0 +1,22 @@
+  try {
+    milliseconds = creationDate * 1000;
+  } catch (error) {
+    return new WrongTimestamp(`Could not convert ${creationDate} to milliseconds`);
```
Hi @mario-nt I think this line is never reached. If I'm not wrong `miliseconds` would be `Infinity` but ti will not throw an exception.

```javascript
// Online Javascript Editor for free
// Write, Edit and Run your Javascript code using JS Online Compiler

let seconds = Number.MAX_VALUE;

let miliseconds = seconds * 1000;
console.log("test1: ", miliseconds); // Infinity

try {
  milliseconds = seconds * 1000;
} catch (error) {
  console.log("error", error);
}
console.log("test2: ", miliseconds); // Infinity
```

josecelano commented on src/helpers/DateConverter.ts:
```diff
@@ -0,0 +1,22 @@
+  } catch (error) {
+    return new InvalidDateError(`Could not create a new date from ${milliseconds}`);
+  }
+  return !isNaN(convertedDate.valueOf()) ? convertedDate.toDateString() : new InvalidDateError(`Could not create a valid date from ${milliseconds}`);
```
Hi @mario-nt In general I prefer positive conditional:

```typescript
  return isNaN(convertedDate.valueOf())
    ? new InvalidDateError(
        `Could not create a valid date from ${milliseconds}`
      )
    : convertedDate.toDateString();
```

See https://softwareengineering.stackexchange.com/a/350474/430081

josecelano commented on src/helpers/DateConverter.ts:
```diff
@@ -0,0 +1,22 @@
+    convertedDate = new Date(milliseconds);
+  } catch (error) {
+    return new InvalidDateError(`Could not create a new date from ${milliseconds}`);
+  }
```
Hi @mario-nt I think the `<Date` function does not thorw any exception:

```javascript
let seconds = Number.MAX_VALUE;

let miliseconds = seconds * 1000;
console.log("test1: ", miliseconds); // Infinity

try {
  milliseconds = seconds * 1000;
} catch (error) {
  console.log("overflow error", error);
}
console.log("test2: ", miliseconds); // Infinity

try {
  convertedDate = new Date(seconds * 1000);
} catch (error) {
  console.log("Date error", error);
}
console.log("test3: ", convertedDate); // Infinity
```

josecelano commented on src/helpers/DateConverter.ts:
Some tests I did:

```TypeScripty
let seconds = Number.MAX_VALUE;

let miliseconds = seconds * 1000;
console.log("test1: ", miliseconds); // Infinity

try {
  milliseconds = seconds * 1000;
} catch (error) {
  console.log("overflow error", error);
}
console.log("test2: ", miliseconds); // Infinity

try {
  convertedDate = new Date(seconds * 1000);
} catch (error) {
  console.log("Date error", error);
}
console.log("test3: ", convertedDate); // Invalid Date

convertedDate = new Date(1701688171 * 1000);
console.log("test4: ", convertedDate);
console.log("test5: ", convertedDate.toISOString());
console.log("test6: ", convertedDate.toUTCString());
```

josecelano replied:
Hi @mario-nt I've just created [another example](https://www.typescriptlang.org/play?ssl=32&ssc=1&pln=1&pc=1#code/C4TwDgpgBAKglgWwgZ2AQwWKBeKA7AVwQCMIAnAbgChRIoAxAezITWGAgBMARN6XVGTh4A5tSoBjADZpkyKAEk8ANzRS4PPgFEyZZlAgAPDnk7ydeslADeAXypUA9ACpnVZ7DQBrFFGAALaE4+KGEoZAgJRlN5ADM9BCgAVTw4QygtMEYJf1hEaDRTKCiVcmB5OGA-Rih-IkKoMgg0YOIpaFjmVmAAOnd3Bn0jDDB2gC4oAEYAdgAGSYA2AA4lgBYAVkmoAFoAPigAIgBZaKhuSKhZ1agAJlmbgGYD90cqWII8CWA4U86WNngSFQIwAFBImmwfnhASh0JgJjDgZgAJQTJj-dhcXgcKAAHwyun01ioUCg7SqwRxuDwEAA7mc+GCId9ooi4VgPJNZtzkeJSU1gAQyHhQsgAHJoMUgykQHqqKQECAAeViIORyJJpKgAH58HTFCo1BpsRALMwQZqtVAAAZKeUaKAynpQADCjAIUk4+EYVXBzRxaD19JlUHijES3yB7Kg8sVEwAJNY-ZDWfkkWBbNbLVANVaJk7gIwTQBlYBCURq6j2ByORxQABqRq9MuQkmiyEY7R6UkYIhBBxgsMmYwOABpQ10AWn2SCZvNlmtNuqKFBa4cTiLzhJLtc7o9niUO12e32B7CbiPx39umzQXPFisNpMegBOZeruvHU5bne3e5PGs6ztJtHT4VsnDrE0oEYZRyFiHtaTbPAj1lE9+0HVAHkvCcMVvTAQTFIhSDIHojgAQQADQAfXrMiABkki0d81wOYD1GbPhnTdD0vTwH1imZAog1AnEwwjacRhjNQ4ymHpphfaYFhfB5JgeVYlgWR5JnWaYIAAageWYlkOCCGRxYR7S9WgCjIEQiAgPBgCQlDu17dDYVWbDrynKNQQOPiqkIEhyAOZjPzYh0nVdd1PW9X1BKgQMaWDEIxL8CTMCkhUIAmAL8CI8gTKAA).

Why do you want to differentiate between an invalid timestamp in the database and a timestamp greater than the maximum Date? Now we are showing the value causing the error so we can find out later what happened.

mario-nt committed:
refactor: [#297]  code cleanup

mario-nt commented:
@josecelano Changes applied.

josecelano commented:
> @josecelano Changes applied.

Hi @mario-nt it looks good but it needs to be rebased.

mario-nt commented:
@josecelano Rebased.

josecelano commented:
ACK 4281526961d821f6a871765c3938ac2144e64311
